### ISSUE TITLE
InvalidOperationException: Did not find any initialized language workers

### ISSUE THREAD

**Issue by AsgerPilegaard** [Azure/azure-functions-host#7618]
This issue can be seen as a continuation of https://github.com/Azure/azure-functions-host/issues/6931.

Message: `Exception while executing function: Functions.keepAlive Did not find any initialized language workers`
Failed Method: `Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcFunctionInvocationDispatcherLoadBalancer.GetLanguageWorkerChannel`

  The exception is seen in our http triggered, timer triggered and service bus triggered azure functions running on consumption plan written in java running on azure function service plans with Windows as operating system.  

#### Occurrence count
  We are currently seeing 600 occurrences of the exception per hour, in a system running a spiked workload of 20 jobs an hour. RetryCount of 6-1 multiplied by servicebus deliveryCount 3 multiplied by the number of azure functions in the job chain 10, allows for a worst case scenario of 3000 occurrences of the exception when running 20 jobs.
  Before the 2nd of July, the occurrence was higher, around 2000 per hour.

#### Inconsistent bug
  The exception does not seem to be consistently occurring on instances where the exception has previously occurred.

#### First occurrence time
14 of May 2021

## Investigative information
Event time | 8/15/2021, 8:31:41 PM (Local time)
Message | Exception while executing function: Functions.keepAlive Did not find any initialized language workers
Exception type | System.InvalidOperationException
Failed method | Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcFunctionInvocationDispatcherLoadBalancer.GetLanguageWorkerChannel
HostInstanceId | d6a3afbd-6481-4124-ba5f-f2150ea86faa
InvocationId | 889c1000-980c-4676-b32d-ac3b72885a58
prop__functionName | Functions.keepAlive

## Repro steps
A. Deploy an http or service bus or Timer triggered azure function in a function app running in consumption plan with the configurations specified in the following section.
B. Load test the azure function to cause a scale out.
C. Language worker initialisation exception is thrown.

#### Function App Service Plan
Operating System: Windows
Location: West Europe 
Zone redundant: Disabled
Status: Ready

#### Function App Details - Site Config
 "linuxFxVersion": "",
"javaVersion": "1.8",
"location": "West Europe",
 "minTlsVersion": "1.2",
 "netFrameworkVersion": "v4.0",
 "maxNumberOfWorkers": null,
 "numberOfWorkers": 1,
"phpVersion": "5.6",
“powerShellVersion": "",
“preWarmedInstanceCount": 0,
"pythonVersion": "",
(Revealed by running az functionapp show -g prodb-publicapi -n prodb-publicapi)

## Attempted fixes
Specifying the java-version to 1.8, was previously unspecified (As specified on the 6931 issue).
Migrate to a new scale unit on the Function App (done by an azure supporter currently only on a non failing function app).

## Related issues
This issue can be seen as a continuation of https://github.com/Azure/azure-functions-host/issues/6931.
https://github.com/Azure/azure-functions-host/issues/6936
https://github.com/Azure/azure-functions-host/issues/5367
https://github.com/Azure/azure-functions-host/issues/5233

### Microsoft Employees that has already verified and worked with the issue, the recent supporter at the top of the list
Ritika Singhal
Luis Ramirez
Harshita Bajpai
@FinVamp1 Finbar Ryan

### The following conclusion was written by an Azure App Service Support Engineer on the 28th of June 
Summary
In the occasions where the Failed to Initialize Language Worker is being logged what we see is:
The Azure Functions Application is started on a new worker
The Functions Host starts
It sends a request out to start the Java Language worker process.
This process starts but never replies back to the Functions host and so the Language Worker process start request times out.
This process continues to run and then shuts down about 30 minutes later.
 
Next Steps 
The start of the Language worker process should initialize the Functions Java Language Worker process and then reply back to the Functions Host.
As this is not happening for a lot of customers we would like your help in understanding the specifics about your application.

**Comment by FinVamp1** [comment]
As discussed in discussions with the issue owner via email.

**Issue Summary**

Your Azure Functions apps implemented with Java will intermittently get into an invalid state where each Function invocation will result in an error “Did not find any initialized language workers”.
This issue happens just after the Functions start-up and will persist on the worker machine hosting that Azure Function app until a restart is issued. 
As your Function Apps are running on the Consumption plan there is no way to implement a restart on individual instances. 

**Issue Investigation and planned solutions from Azure Functions Host.**

As I think I mentioned previously the timeline for what happens looks like this.

1)	The Function App is allocated a new instance. 
2)	The Functions Runtime is initialized
3)	It detects that this is a Function app and starts to Initialize the Java Language worker.
4)	The Java Process starts but never completes initialization.
5)	The Functions Runtime detects that the worker channel has not started and logs the error message on further invocations. 

The engineering team have investigated this issue and there are currently two channels to the investigation.

1)	The timeout for the language worker initialization may need to be adjusted and the team are looking at adding additional logging. This logging will help identify the average startup time and how many times this average fails just below the timeout.
If it does then increasing the timeout maybe the solution here.
2)	The other approach is to ensure that we don’t accept invocations on an Azure Functions Application that is in this state and potentially restart the language worker if it is in this state.

Item 1 is still being discussed and a work item has not been created to track this yet. I will follow up on this today. 
Item 2 is being tracked under this work item(Erroneous dispatcher state when webhostLanguageWorkerProcess doesn't come up · Issue #7590 · Azure/azure-functions-host (github.com))and a fix has been identified. This fix should roll out in the next version of the Azure Functions runtime. These releases are generally ever 30 to 60 days.

**Comment by biltongza** [comment]
I'm getting the same issue with dotnet-isolated functions. The first invocation of my timer functions works as expected, but every invocation after that fails with "Did not find any initialized language workers" and times out after ~300 seconds.

**Comment by biltongza** [comment]
#7558 seems related.

**Comment by AsgerPilegaard** [comment]
Hi Finbar.
Thank you for your acknowledgement.

FEEDBACK
1. Where can we follow the work item about monitoring the startup issues?
2. We were notified about the release of the fix for issue #7590, several months ago. When exactly can we expect the the fix to be released? (We have the system in production and need a reliable and specific solution plan to present to our stakeholders)

STATUS
The bug is still very much visible, only in the past day we have seen 7,543 occurrences.
Application Insights Log Query: 
`exceptions | where timestamp >= ago(1d)
| where innermostMessage has "Did not find any initialized language workers"´

FURTHER INVESTIGATION
The app setting FUNCTIONS_WORKER_PROCESS_COUNT specifies the number language worker processes  (https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings). Is it correct that the function runtime throws the exception when the java language worker process startup never completes? 
Would increasing the FUNCTIONS_WORKER_PROCESS_COUNT increase the number of initialized language workers? Would these be initialized in parallel? Could this potentially increase the change of having an initialized and usable language worker on the instance? 
What would be the overhead and downside to increasing this app setting? Our service bus triggered azure function instances would start running more jobs on the same instance if the maxConcurrentCalls service bus setting is high enough.

**Comment by FinVamp1** [comment]
Hello @AsgerPilegaard , thanks for following up on this . I have an alert for your applications and when it happens I am going to try and investigate more.  I am following up on [the fix work item](https://github.com/Azure/azure-functions-host/issues/7590) to see when it is going to be released. The FUNCTIONS_WORKER_PROCESS_COUNT could help and I will look into this a bit more today.  It is correct about when you see this error is means that the language worker never completed intialization. The Worker Process Count doesn't start all of the Worker Processes at Startup. Instead it creates new ones every 10 seconds as needed. "Language worker processes are spawned every 10 seconds until the count set by FUNCTIONS_WORKER_PROCESS_COUNT is reached. " Can you perhaps start testing with 4 and see if it helps? For the Consumption plan you are limited to 1 processor core and so if your workload is CPU bound I wouldn't suggest going directly to 10. I'll send this update on your support incident as well.

**Comment by AsgerPilegaard** [comment]
Hello Finbar.

Thank you for your explanation.
What is the status on
1. Where can we follow the work item about monitoring the startup issues?
2. We were notified about the release of the fix for issue Erroneous dispatcher state when webhostLanguageWorkerProcess doesn't come up #7590, several months ago. When exactly can we expect the the fix to be released? (We have the system in production and need a reliable and specific solution plan to present to our stakeholders)

Using multiple language workers would only mitigate the issue, and not remove it. Furthermore we have seen some timeout issues when running an azure function in python. The service bus parameter maxConcurrentCalls defaults to 16 and causes the azure function load balancer to distribute multiple messages to a singe language worker (since the possible 10 language workers does not boot up at start up but at 10 sec intervals), causing the service bus lock to timeout. So it is important that we have the freedom to adjust the scaling parameters to optimize our scale out. Therefore we currently have not assigned any one to attempt the FUNCTIONS_WORKER_PROCESS_COUNT mitigation fix.

We have decided to attempt to change all our serverfarms to run on linux servers instead of windows servers. It seems like this Language Worker Initialization issue only occurs when running on windows servers. 

For those who might have the same issue, our ARM templates look like this:
`{
      "type": "Microsoft.Web/serverfarms",
      "apiVersion": "2020-09-01",
      "name": "[variables('serverfarmName')]",
      "location": "[resourceGroup().location]",
      "kind": "app,linux",
      "sku": {
        "name": "Y1",
        "tier": "Dynamic"
      },
      "properties": {
        "name": "[variables('serverfarmName')]",
        "computeMode": "Dynamic",
        "reserved": true
      }
    }`
    
`{
      "type": "Microsoft.Web/sites",
      "apiVersion": "2018-11-01",
      "name": "[variables('functionAppName')]",
      "location": "[resourceGroup().location]",
      "kind": "functionapp,linux",
      "tags": "[variables('tags')]",
      "identity": {
        "type": "UserAssigned",
        "userAssignedIdentities": {
          "[parameters('myFunctionAppMsiResourceId')]": {}
        }
      },
      "dependsOn": [
        "[resourceId('Microsoft.Web/serverfarms', variables('serverfarmName'))]",
        "[resourceId('Microsoft.Storage/storageAccounts', variables('storageAccountName'))]"
      ],
      "properties": {
        "httpsOnly": "true",
        "serverFarmId": "[resourceId('Microsoft.Web/serverfarms', variables('serverfarmName'))]",
        "siteConfig": {
          "linuxFxVersion": "Java|8"
        }
      }
}`

**Comment by FinVamp1** [comment]
Hello,

@AsgerPilegaard  the fix for the Language Worker initialization is in[https://github.com/Azure/azure-functions-host/issues/7590]( https://github.com/Azure/azure-functions-host/issues/7590 )  and it has been checked in. This did not make into the most recent release 3.1.2 but should be in the one after this. 

The monitoring startup times doesn't have a work item assigned but @vrdmr  and @Francisco-Gamino  have been following up on that.

Thanks for the notes on the Functions Worker Process Count. It will be interesting to see your results when running on Linux workers. 

@Francisco-Gamino  can help monitor this issue while I am away for the next couple of weeks.

**Comment by AsgerPilegaard** [comment]
Hello Finbar.

Changing to run azure functions on linux servers has eliminated the InvalidOperationException completely. However I would assume that this task is still very much an issue on windows. 
Ritika Singhal informed me that the fix #7590 also did not make it into release 3.2.0, but supposedly will be part of 3.3.0, and should complete in 1-2 weeks if not blocking issues are encountered. I am eager to hear the status on this. 
I do not think microsoft should consider this task closed, before microsoft has been verified that the fix actually works.

Changing to run our azure functions on linux servers, has however caused the application insights trace and dependency map to be lost. I assume that the windows runtime has an agent that intercepts all http requests, which uploads the request and response metadata to the application insights service and add the traceparent as a header to the http requests to accommodate the W3C protocol, enabling a coupled traced across http requests. Is it correct that such an agent exist on windows and is not implemented for linux?  When we attempt to upload the dependencies or traces to application insights nothing happens #7688. Any help would be very much appreciated. 

What is the recommended setup or action from microsoft in our scenario? We now have a system running on linux servers with significantly fewer exceptions. However the dependency map is gone, and we have to apply the W3C protocol headers ourselves in order to regain the trace in application insights.

**Comment by amamounelsayed** [comment]
@AsgerPilegaard  Thank you so much for your message. Please can you provide how do you send the data to the AI from Azure Functions?

Are you using Distributed-Tracing-for-Java-Azure-Functions?
https://github.com/Azure/azure-functions-java-worker/wiki/Distributed-Tracing-for-Java-Azure-Functions

Thank you so much,
Ahmed

**Comment by AndersDyhr** [comment]
@amamounelsayed thanks for your interest in our AI issue. The available information is in issue [#7688](https://github.com/Azure/azure-functions-host/issues/7688). Would it be possible for you to post in that issue, to keep the two issues seperated?

**Comment by AndersDyhr** [comment]
@FinVamp1 we can see that 3.3.1 is now released ([link](https://github.com/Azure/azure-functions-host/releases)), has the fix made it into this release?

Furthermore, when we set the FUNCTIONS_EXTENSION_VERSION to either 3.3.1 or 3.3.0 it still selects 3.2.0, so it seems that 3.3.1 is stil no available to us even though it is on the release page?

**Comment by rosidigital** [comment]
@AsgerPilegaard I have run into the same problem, both on windows and linux. The most annonying thing is that it comes out of the blue. We have a QA environment and a PRODUCTION environment - just compared the ARM templates and they are exactly identical - but somehow the PRODUCTION will not start! 

Even tried creating a new Azure Function and deploying to that - even that didn't help. 

Will probably dish Azure Functions (for the 2. time in my short life) and just create an ordinary webapp with some HangFire. :-(

**Comment by amamounelsayed** [comment]
@rosidigital can you share your app language and type (dedicated, premium, consumption) please?

**Comment by rosidigital** [comment]
@amamounelsayed c# (.net 5) / consumption.

**Comment by amamounelsayed** [comment]
Thank you @rosidigital, we are currently investigating the issue, we will keep the thread update it

**Comment by sashuprosperoware** [comment]
Any update guys, we are getting this error on Java Function App as well on invocation from message from service bus queue.

**Comment by rosidigital** [comment]
I ended up dumping my azure functions for a hangfire setup for our purpose it makes just as much sense.

**Comment by biltongza** [comment]
I switched from a Linux consumption plan to a Windows consumption plan, and dotnet-isolated just started working without any issues. I tried getting the Linux one to work for *days* because of this issue.

**Comment by rosidigital** [comment]
@biltongza i don't think that it was the switch that fixed it - its probably just because it was a new service or plan. Two months ago I experienced the same, it started working when I switched to Linux - but then suddenly out of the blue it stopped working again.

**Comment by biltongza** [comment]
I tried making a new Linux consumption plan from scratch as well, and had the same issue where it didn't find any initialized workers and died. Switching to windows was my last ditch effort and it just worked magically.

**Comment by amamounelsayed** [comment]
@rosidigital  @biltongza  could you please share the app names and please can you share the language?
@sashuprosperoware could you please share the app name? 
Thank you so much

**Comment by richsage** [comment]
Not sure if it helps the diagnosis, but I saw this error appear with effect from _1 November 2021_. All working absolutely fine day before. No code / infrastructure changes at all for a week or so beforehand.

.NET 5 (isolated) Linux function app, Linux consumption app service plan

**Comment by fabiocav** [comment]
@richsage would you mind opening a separate issue sharing a timestamp and either your application name or an invocation ID + region so we can take a closer look? 
Thanks!

**Comment by richsage** [comment]
@fabiocav Started writing the issue, went to get invocation ID and it turns out things started working again randomly overnight...! Nothing touched from our side. Perhaps an update within the platform itself? North Europe region.

![image](https://user-images.githubusercontent.com/231551/140609668-0dcb1b57-a5c6-4104-bb38-1c3a7413603e.png)

I won't continue to open the issue but will keep an eye on things and will open if it fails again :)

**Comment by liliankasem** [comment]
> @fabiocav Started writing the issue, went to get invocation ID and it turns out things started working again randomly overnight...! Nothing touched from our side. Perhaps an update within the platform itself? North Europe region.
> 
> ![image](https://user-images.githubusercontent.com/231551/140609668-0dcb1b57-a5c6-4104-bb38-1c3a7413603e.png)
> 
> I won't continue to open the issue but will keep an eye on things and will open if it fails again :)

Good to know it's working, but I think we'd still like to investigate  as we're seeing similar issues with others. Do you mind still creating that issue?

**Comment by mattosaurus** [comment]
In my case I got this error when referencing an enviroment variable in Program.cs that wasn't present in the Funcation app configuration settings.

I'm guessing this exception is thrown whenever the host fails to start so there's probably a lot of potential causes.

**Comment by richsage** [comment]
@liliankasem thanks, have opened #7843, let me know if you need more details :)

**Comment by liliankasem** [comment]
We've been making frequent improvements as we identify scenarios that lead to this error state (which is, unfortunately, quite generic). The best approach here is to open a support ticket (or new issue) any time you run into this message so we can take a closer look at the specific case.

Closing this issue to promote that approach, but please let us know if you have any additional questions.
